docs/nsm_api.md

```diff
@@ -347,7 +347,148 @@ message ConnectionContext {
+- A Network Service Client (NSC) which is connected to a local mechanism 'memif' is cross conneted to a Network Service Endpoint (NSE) via a local mechanism 'memif' if they are both on the same Node.  This cross connect can be thought of as a pair ('memif','memif').
+- A Network Service Client (NSC) which is connected to a local mechanism 'memif' is cross connected to a Network Service Endpoint (NS) via the local mechanism 'kernel interface'.  This cross connect can be thought of as the pair ('memif','kernel interface')
+
+```proto
+
+service NSM2NSMD {
+    rpc CreateCrossConnnect(CreateCrossConnectRequest) returns CreateCrossConnectReply;
+    rpc UpdateCrossConnect(UpdateCrossConnectRequest) returns UpdateCrossConnectReply;
+    rpc DeleteCrossConnect(DeleteCrossConnectRequest) returns DeleteCrossConnectRequest;
+    rpc ListAndwatchCrossConnects(ListAndWatchCrossConnectRequest) returns (stream ListAndWatchCrossConnectReply)
+    rpc ListAndWatchCrossConnectMechanisms(ListAndWatchCrossConnectMechanismsRequest) returns (stream ListAndWatchCrossConnectMechanismsReply)
+}
+
+service NSMD2NSM {
+    rpc Register(NsmdRegistrationRequest) return NsmDRegistrationResponse
+}
+
+/*
+ * CreateCrossConnectRequest - Message to create a a CrossConnect
+ *
+ * id - id of the cross connect.  Must be unique in the context of this NSM and NSMD.
+ *      id is selected by the NSM.
+ * mechanism1 - The LocalMechanism to connect to a local NSC or NSE.
+ *              The presumption is that all cross connects involve at least one
+ *              Pod local to the Node.
+ * mechanism2 - The other end of the cross connect.  May be either a local mechanism
+ *              if cross connecting two Pods on the same Node, or a remote mechanism
+ *              if cross connecting to something remote from the Node.  Note:
+ *              RemoteMechanism here is the same message as used in NSM2NSM API.
+ */ 
+message CreateCrossConnectRequest {
+    string id = 1;
+    LocalMechanism mechanism1 = 2;
+    oneof mechanism2 {
+        RemoteMechanism remoteMechanism = 3;
+        LocalMechanism localMechanism = 4;
+    }
+}
+
+message UpdateCrossConnectRequest {
+    string id = 1;
+    LocalMechanism mechanism1 = 2;
+    oneof mechanism2 {
+        RemoteMechanism remoteMechanism = 3;
+        LocalMechanism localMechanism = 4;
+    }
+}
+
+message DeleteCrossConnectRequest {
+    string id = 1;
+}
+
+message LocalMechanism {
+    LocalMechanismType type = 1;
+    oneof parameters {
+        KernelInterfaceParameters = 1;
+        MemifParameters = 2;
+        VhostUserParameters = 3;
+    }
+}
+
+enum LocalMechanismType {
+    KERNEL_INTERFACE = 0;
+    MEMIF = 1;
+    VHOST_USER = 2;
+}
+
+message KernelInterfaceParameters {
```

Comment:
I think we should not have per type messages, we should use map[string]string, in this case kernel interface will have two properties and the code dealing with kernel interface will expect to find them in interface parameters.